src/coreclr/jit/lowerxarch.cpp

```diff
@@ -130,6 +130,38 @@ void Lowering::LowerStoreIndir(GenTreeIndir* node)
             return;
         }
     }
+    else if (node->AsStoreInd()->Data()->OperIs(GT_CNS_DBL))
+    {
+        // Optimize *x = DCON to *x = ICON which is slightly faster on xarch
+        GenTree*  data   = node->AsStoreInd()->Data();
+        double    dblCns = data->AsDblCon()->gtDconVal;
+        ssize_t   intCns = 0;
+        var_types type   = TYP_UNKNOWN;
+
+        if (node->TypeIs(TYP_FLOAT))
+        {
+            float fltCns = static_cast<float>(dblCns); // should be a safe round-trip
+            intCns       = static_cast<ssize_t>(*reinterpret_cast<UINT32*>(&fltCns));
+            type         = TYP_UINT;
+        }
+#ifdef TARGET_AMD64
+        else
+        {
+            assert(node->TypeIs(TYP_DOUBLE));
```

Comment:
Are we not concerned about this optimization on x86? I'd imagine a decomposed store is still faster than reading the constant from memory and then storing it (should also be smaller codegen)

Reply:
> should also be smaller codegen

Not really since we store FP constants in the data section
so it's 2*6=12 bytes of two `mov`'s vs 8+4=12 bytes of two `vmovsd`.
Kind of makes sense to freed the data section but it won't just work if I remove that #ifdef and I'll have to split the constant myself into two stores here (complicates the code)

Reply:
> Kind of makes sense to freed the data section but it won't just work if I remove that #ifdef and I'll have to split the constant myself into two stores here (complicates the code)

I think on x86 you would need to change `GT_CNS_DBL` node type to `GT_CNS_LNG` (not to `GT_CNS_INT` as on x64) but it shouldn't require two separate stores.

Reply:
Discussed offline with Egor and decided to leave as is.